lib/spack/spack/environment/cache_shell_script.py

```diff
@@ -0,0 +1,115 @@
+# Copyright Spack Project Developers. See COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+import os
+import sys
+from datetime import datetime
+
+import spack.environment.shell
+from spack.util.environment import EnvironmentModifications
+
+
+def path_to_env_activate_shell_script(env, shell):
+    """Returns to path to the shell script to activate the specified env for the shell that
+    the user is running
+
+    Args:
+        env: The environment whose shell script we are returning the path of
+        shell: The shell that the user is running on
+    """
+
+    return os.path.join(env.path, ".spack-env", f"activate.{shell}")
+
+
+def path_to_env_deactivate_shell_script(env, shell):
+    """Returns to path to the shell script to activate the specified env for the shell that
+    the user is running
+
+    Args:
+        env: The environment whose shell script we are returning the path of
+        shell: The shell that the user is running on
+    """
+
+    return os.path.join(env.path, ".spack-env", f"deactivate.{shell}")
+
+
+def write_env_activate_script(env, view):
+    """This does a thing
+
+    Args:
+    """
+
+    shells_avail = ["sh", "csh", "fish"]
+
+    if sys.platform == "win32":
+        shells_avail.extend(["bat", "pwsh"])
+
+    for shell in shells_avail:
+        cmds = spack.environment.shell.activate_header(env, shell, view=view)
+
+        env_mods = EnvironmentModifications()
+
+        cmds += env_mods.shell_modifications(shell)
+
+        activate_script_path = os.path.join(env.path, ".spack-env", f"activate.{shell}")
+        with open(activate_script_path, "w", encoding="utf-8") as f:
+            f.write(
+                f"### Script created by spack (https://github.com/spack/spack) {datetime.today().strftime('%Y-%m-%d')}\n\n"
+            )
+            f.write(cmds)
+
+
+def append_env_activate_script(env, shell, prompt, view):
+    """This does a thing
+
+    Args:
+    """
+
+    env_mods = EnvironmentModifications()
+    env_mods.extend(spack.environment.shell.activate(env=env, view=view))
+
+    activate_cmds = env_mods.shell_modifications(shell)
+
+    prompt_cmds = spack.environment.shell.activate_with_prompt(shell, prompt)
+    view_cmds = spack.environment.shell.activate_with_view(shell, view)
+
+    activate_script_path = os.path.join(env.path, ".spack-env", f"activate.{shell}")
+
+    with open(activate_script_path, "r", encoding="utf-8") as file:
+        if activate_cmds in file.read():
+            activate_cmds = ""
+        if prompt_cmds in file.read():
+            prompt_cmds = ""
+        if view_cmds in file.read():
```

Comment:
The `read()` at line 79 will read all of the contents of the file at once while those at 81 and 83 will result in empty strings.  See the docs at https://docs.python.org/3/tutorial/inputoutput.html#methods-of-file-objects.